### DISCUSSION TITLE
esp32 interrupt timer problem

### DISCUSSION THREAD

**Discussion by bristol406** [micropython/micropython#9926] [Using MicroPython]
I am running an esp32 with a hardware watchdog timer which performs a reset if the watchdog board doesn't receive a pulse from the program within 30s.  I am using timer 0 to kick the watchdog every 10s by sending a pulse on GPIO pin. Everything seemed to be fine initially but I discovered that when the program fails the interrupt timer is still running and the watchdog is still receiving  the keepalives. Is there any way around this or an alternative solution to my problem ?

**Comment by karfas** [comment]
You need to feed the watchdog from some **periodic function within your program**. There is no short way (like the timer) around this.

> when the program fails

Note also that a watchdog is also not a replacement for proper error handling. Most, if not all "When the program fails" (whatever this means) conditions should be handled by your software.

**Reply by bristol406** [reply]
Thanks karfas. I was originally feeding the watchdog within my program but that was an issue. I found that the urequests library function which has no timeout parameter was occasionally waiting/hanging for a period exceeding my 30s watchdog timer so esp32 was being reset.  This is why I tried using esp32 timer. I believe that I handle all possible errors using "try - Except" statements. My diagnosis citing urequests could be wrong - I can't see why it would wait that length of time.

**Reply by karfas** [reply]
It is clear this needs to be fixed in asyncio and/or the socket library.
On the other hand, it just proves my point.
The watchdog will protect you against unforeseeable lock-ups. 
Delegating the Watchdog trigger to an interrupt-driven function completely unrelated to your program contradicts the purpose of the watchdog, isn't it ?
